tidy3d/components/medium.py

```diff
@@ -615,6 +615,56 @@ def _interp(
         interp_dataset = interp_dataset.where(interp_dataset <= max_val, max_val)
         return interp_dataset
 
+    def grids(self, bounds: Bound) -> Dict[str, Grid]:
+        """Make a :class:`.Grid` corresponding to the data in each ``eps_ii`` component.
+        The min and max coordinates along each dimension are bounded by ``bounds``."""
+
+        rmin, rmax = bounds
+        pt_mins = dict(zip("xyz", rmin))
+        pt_maxs = dict(zip("xyz", rmax))
+
+        def make_grid(scalar_field: ScalarFieldDataArray) -> Grid:
+            """Make a grid for a single dataset."""
+
+            def make_bound_coords(coords: np.ndarray, pt_min: float, pt_max: float) -> List[float]:
+                """Convert user supplied coords into boundary coords to use in :class:`.Grid`."""
+
+                # get coordinates of the bondaries halfway between user-supplied data
+                coord_bounds = (coords[1:] + coords[:1]) / 2.0
```

Comment:
I don't understand this, it looks like you're taking the cell centers? But then you're passing this to a Grid construction, which normally takes boundaries?

Reply:
The `coords` store the centers of the cells in the `eps_dataset`, so the boundaries are the halfway points. Am I missing something?

Reply:
Ok, I see what you mean, I think my confusion here was that you're not making a single grid for all three components, but instead an individual grid for every component, where you assume that the values are defined at the grid center.

The only point here is that in the case of a non-uniform grid, grid centers are the halfway points between grid boundaries, but grid boundaries are not the halfway points between grid centers. Consider these two cells: |--x--|---x---|

Separately, isn't there potentially an issue of not covering the full CustomMedium domain since the array of boundaries has one fewer element than the input array of coordinates (the span is missing half a step on each side)?

Reply:
On the first point about nonuniform, seems I did neglect that, I think it's ok for right now but any suggestions on how to fix it?

On second point, later on I add the bounds for the intersection of the geometry and simulation to the boundary coords list, so it takes care of that missing step on each side.

Reply:
> On the first point about nonuniform, seems I did neglect that, I think it's ok for right now but any suggestions on how to fix it?

Not really, I think there's a piece of information missing. You need to know the size of one of the cells and then the rest are uniquely defined in a domino effect.